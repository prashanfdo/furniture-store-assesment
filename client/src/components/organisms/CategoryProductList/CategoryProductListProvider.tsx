import constate from "constate";
import useCategoryProductListQuery from "queries/useCategoryProductListQuery";

type CategoryProductListProviderProps = {
  categoryId: string;
};
const [CategoryProductListProvider, useCategoryProductList] = constate(
  ({ categoryId }: CategoryProductListProviderProps) => {
    const { loading, error, data } = useCategoryProductListQuery(categoryId);
    return {
      loading,
      error,
      data: data?.categories?.[0],
    };
  },
);

export { useCategoryProductList };

export default CategoryProductListProvider;
