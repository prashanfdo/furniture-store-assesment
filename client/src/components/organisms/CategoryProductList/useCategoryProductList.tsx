import useCategoryProductListQuery from "queries/useCategoryProductListQuery";

const useCategoryProductList = (categoryId: string) => {
  const { loading, error, data } = useCategoryProductListQuery(categoryId);
  return {
    loading,
    error,
    data: data?.categories?.[0],
  };
};

export default useCategoryProductList;
