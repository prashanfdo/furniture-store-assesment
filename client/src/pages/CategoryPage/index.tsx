import ProductList from "components/organisms/ProductList";
import React from "react";
import { useMatches } from "react-router-dom";

export default function CategoryPage() {
  const matches = useMatches();

  const categoryId = matches?.[0]?.params?.category || "156126";

  return (
    <>
      <ProductList categoryId={categoryId} />
    </>
  );
}
