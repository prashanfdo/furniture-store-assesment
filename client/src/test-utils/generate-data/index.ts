export * from "./article";
export * from "./category";
export * from "./search";
