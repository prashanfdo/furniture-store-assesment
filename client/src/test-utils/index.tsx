import { render, RenderOptions, renderHook, RenderHookOptions, RenderHookResult } from "@testing-library/react";
import fs from "fs";
import React, { FC, ReactElement } from "react";

import user from "@testing-library/user-event";
import { ProvidersForTests } from "providers";

type GraphQlMocks = React.ComponentProps<typeof ProvidersForTests>["mocks"];

type WrapperProps = React.ComponentProps<typeof ProvidersForTests> & {
  children: React.ReactNode;
};
const Wrapper: FC<WrapperProps> = ({ children, mocks, initialRoute }: WrapperProps) => {
  return (
    <ProvidersForTests mocks={mocks} initialRoute={initialRoute}>
      {children}
    </ProvidersForTests>
  );
};
const wrapper =
  (mocks: GraphQlMocks, initialRoute?: string) =>
  ({ children }: { children: React.ReactNode }) =>
    (
      <Wrapper mocks={mocks} initialRoute={initialRoute}>
        {children}
      </Wrapper>
    );

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper"> & { gqlMocks: GraphQlMocks; initialRoute: string },
) => {
  const view = render(ui, { wrapper: wrapper(options?.gqlMocks, options?.initialRoute), ...options });
  const style = document.createElement("style");
  style.innerHTML = fs.readFileSync("src/index.out.css", "utf8");
  document.head.appendChild(style);

  return view;
};

export const customRenderHook = <T, R>(
  callback: (props: T) => R,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { gqlMocks?: GraphQlMocks; initialRoute?: string },
): RenderHookResult<R, T> => {
  return renderHook(callback, { ...options, wrapper: wrapper(options?.gqlMocks, options?.initialRoute) });
};

export { faker } from "@faker-js/faker";
export * from "@testing-library/react";
export * from "./generate-data";
export { customRender as render };
export { customRenderHook as renderHook };
export { user };
