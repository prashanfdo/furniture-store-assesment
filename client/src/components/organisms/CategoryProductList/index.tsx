import { ArticleListHeader } from "components/atoms";
import { ArticleGrid, SEO } from "components/molecules";
import useCategoryProductList from "./useCategoryProductList";

type CategoryProductListProps = {
  categoryId: string;
};
const CategoryProductList: React.FC<CategoryProductListProps> = ({ categoryId }: CategoryProductListProps) => {
  const { data, loading, error } = useCategoryProductList(categoryId);

  return (
    <>
      <div className="max-w-[100rem] mx-auto grid grid-rows-[auto_1fr] py-2 px-8">
        {error && <div>Error Occurred. Please try refreshing the page.</div>}
        {(data || loading) && (
          <>
            <SEO
              title={`${data?.name} Category`}
              description={`Category with ${data?.articleCount} products`}
              name={data?.name || "Category"}
              type="products"
            />
            <ArticleListHeader name={data?.name} articleCount={data?.articleCount} loading={loading} />
            <ArticleGrid products={data?.categoryArticles.articles || []} loading={loading} />
          </>
        )}
      </div>
    </>
  );
};

export default CategoryProductList;
