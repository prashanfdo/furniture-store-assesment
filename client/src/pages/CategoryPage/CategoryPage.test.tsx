import { render, screen } from "test-utils";
import CategoryPage from "./index";
import CategoryProductList from "components/organisms/CategoryProductList";

jest.mock(
  "components/organisms/CategoryProductList",
  () =>
    ({ categoryId }: React.ComponentProps<typeof CategoryProductList>) =>
      <div data-testid="category-product-list" data-category-id={categoryId} />,
);
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useMatches: () => [{ params: { category: "test" } }],
}));

describe("CategoryPage", () => {
  it("should render", () => {
    const { container } = render(<CategoryPage />);
    expect(container).toMatchSnapshot();
    expect(screen.getByTestId("category-product-list")).toHaveAttribute("data-category-id", "test");
  });

  // TODO: add more tests
});
