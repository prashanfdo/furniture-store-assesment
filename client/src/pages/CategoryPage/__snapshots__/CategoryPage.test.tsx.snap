// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryPage should render 1`] = `
<div>
  <div
    data-category-id="test"
    data-testid="category-product-list"
  />
</div>
`;
